---
title: "Tutorial on `solarius` R package"
author: "Andrey Ziyatdinov, Helena Brunel"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: united
    toc: true
    toc_depth: 3
    number_sections: true
  pdf_document:
    toc: true
    toc_depth: 3
    number_sections: true
bibliography: ref.bib
---

```{r chunk_opt}
opts_chunk$set(eval = TRUE)
```

# Introduction

In human genetics research, the study of a complex disorder is targeted 
to discover its underlying genetic architecture.
Whereas simple Mendelian diseases are results of a mutation in a single gene,
for example [Huntington's disease](http://en.wikipedia.org/wiki/Huntington%27s_disease),
complex disorders or phenotypes are the product of the action of multiple genes and environmental factors
as well as their interactions.
For instance, obesity is a complex disease,
which genetic basis might be better dissected 
via measurable quantitative variation in phenotypes,
such as body-mass index, closely related to the disease risk.

One of the study designs to identify the quantitative trait loci (QTLs)
that affect complex phenotypes is based on examination 
of related individuals grouped by families (pedigrees).
Pedigree-based designs allow for both QTL linkage and association mappings.
In addition, QTL analysis with pedigrees 
possess more power to detect rare variants,
permits explicit control for population stratification in association studies,
and suit for examination of parent-of-origin and sex-specificity effects.

The analysis of the pedigree data is traditionally based on variance component (VC) model, 
also known as linear mixed model (LMM).
`SOLAR` is one of the well-known and widely used tools
that implements the variance component models and other routines needed for the QTL analysis [@Almasy1998]
([http://solar.txbiomedgenetics.org/](http://solar.txbiomedgenetics.org/)).
The name `SOLAR` stands for Sequential Oligogenic Linkage Analysis Routines.
The analyses available in `SOLAR` include 
customized quantitative genetic analysis, 
advanced linkage analysis, 
SNP association analysis (QTN, QTLD, and MGA). 
Other operations are implemented for calculation of marker-specific or multipoint identity-by-descent (IBD) matrices 
in pedigrees of arbitrary size and complexity.
The linkage analysis allows for multiple quantitative traits and/or discrete traits 
which might involve multiple loci (oligogenic analysis), dominance effects, household effects, and interactions.

`solarius` provides an interface from the program `SOLAR` to 
the environment for statistical computing `R`.
On one side, main `SOLAR` procedures can be called in `R` by one or a few commands from the `solarius` package,
that makes the analysis more user-friendly with the same power of computations implemented in `SOLAR`.
On the other side, the `solarius` user benefits from the infrastructure available 
with the R environment and its packages, 
such as interactive data manipulation, data visualization, parallel computation and so on.

Other tools of the interest (similar to `SOLAR`) include `Mendel`, `MMM`, `GEMMA` and `FAST-LMM`,
but the R package `solarius` presented here has been especially developed to work with `SOLAR`.

This vignette is a tutorial on the R package `solarius`.
The document contains a brief description of main statistical models (polygenic, association and linkage) 
implemented in `SOLAR` and accessible via `solarius`, 
installation instructions for both `SOLAR` and `solarius`,
reproducible examples on synthetic data sets available within the `solarius` package.
The examples are intended to cover all the necessary steps 
in the quantitative genetic analysis with `SOLAR`:
data loading, data exploration, running statistical models 
and interpreting the obtained results by printing, summarizing and plotting the models in R.

# Methodology

## Linear mixed model

Consider a particular trait, such as body-mass index, 
is observed in $n$ individuals grouped by families or clusters of related individuals.
A $n \times 1$ vector $y$ contains the phenotypic values measured for the trait.
One assumes that these observations are described adequately by a linear model
with a $p \times 1$ vector of fixed effects ($\beta$)
and a $q \times 1$ vector of random effects ($u$) [@Lynch1998, Chapter 26, pp. 746].
In matrix form,

$$ y = X \beta + Z u + e $$

where $X$ and $Z$ are respectively $n \times p$ and $n \times q$ incidence matrices,
and $e$ is a $n \times 1$ vector of residual deviations
assumed to be distributed independently of the random effects.
Denote the $n \times n$ covariance matrix for the vector $e$ by $R$
and the the $q \times q$ covariance matrix for the vector $u$ by $G$.

The first element of the vector $\beta$ is typically the population mean,
and other elements may be age, gender, diet and so on.
The elements of the incidence matrices are usually equal to 0 or 1,
that indicate the effect contribution to the individuals's phenotype.
The model is referred to as a _mixed model_,
because this model jointly accounts for fixed and random effects.

Giving the two terms of the random effects $u$ and $e$,
the first term typically accounts for the contribution from random genetic effects
such as additive genetic values.
The second term accounts for the residual variance,
and one usually assumes that residual errors have constant variance and are uncorrelated,
so that $R$ is a diagonal matrix, with $R = \sigma_e^2 I$.

Two more equations for the means and variances of the component vectors
of the mixed model can be derived on the basis of its definition.
Since $E(u) = E(e) = 0$ by definition,

$$ E(y) = X \beta $$

Excluding the difference among individuals due to fixed effects 
and calling the assumption that $u$ and $e$ are uncorrelated,
the covariance matrix for the vector of observations $y$ is

$$ Var(y) = V = Z G Z^{T} + R $$

Now the mixed model can be rewritten in a compact form

$$ y = X \beta + Z u + e$$ 

$$
\mbox{ where } \mbox{ } \mbox{ } u \sim (0, G) \mbox{ and } e \sim (0, R) \\
\mbox{ implying } \mbox{ } \mbox{ } y \sim (X \beta, V) = (X \beta, Z G Z^T + R)
$$

For the mixed model,
$y$, $X$ and $Z$ are observed variables,
while $\beta$, $u$, $R$ and $G$ are generally unknown variables.
Thus, the analysis consists of two complementary estimation procedures:
(1) estimation of the covariance matrices $G$ and $R$,
and (2) estimation of the vectors of fixed and random effects, $\beta$ and $u$.
The covariance matrices are usually functions of a few unknown parameters 
or variance components.
For instance, the covariance matrix of the residuals errors
typically has the form $R = \sigma_e^2 I$,
where $\sigma_e^2$ is the parameter to be estimated.

The description of the two estimation procedures is out of scope of this document,
and the reader is referred to [@Lynch1998, Chapter 26-27], 
where the necessary theoretical material and accompanying examples are given.
On the other hand, the implementation of the estimation procedures in `SOLAR`
is discussed elsewhere [@Almasy1998].

The next three sections will present `SOLAR` models,
which follow the general definition of linear mixed model given above,
but parametrization of the models and estimation procedures 
are implemented in a `SOLAR`-specific way.

## Polygenic model in `SOLAR`

The polygenic model is a basis for any quantitative genetic analysis in `SOLAR`.
The model includes fixed effects ($\beta$) 
and three random effects (the genetic additive effect $g$, 
the house-hold effect $c$, and the residual effect $e$).

$$y = X\beta + g + c + e$$

$$\mbox{ where } \mbox{ } \mbox{ } g \sim (0, \sigma_g^2 A) 
  \mbox{ and } \mbox{ } \mbox{ } h \sim (0, \sigma_c^2 H)
  \mbox{ and } \mbox{ } \mbox{ } e \sim (0, \sigma_e^2 I)$$ 
  
$$ \mbox{ implying } \mbox{ } \mbox{ } y \sim (X\beta, V) = (X\beta, \sigma_g^2 A + \sigma_c^2 H + \sigma_e^2 I)$$

The first $g$ term accounts for the additive genetic effect
with the covariance matrix $A = 2 K$, 
where $K$ is the kinship matrix.
`SOLAR` estimates the empirical kinship matrix based on the identifier (ID) fields in the pedigree data,
such as individual ID `id`, gender `sex`, family ID `fam`, father ID `fa`, mother ID `mo`,
and indicator of identical individuals, e.g. monozygotic twins, `mztwin`
[@SOLARDOC, Appendix 1, [file-pedigree](http://helix.nih.gov/Documentation/solar-6.6.2-doc/91.appendix_1_text.html#file-pedigree) command].

The second $c$ term represents an annotated environmental effect.
It is named as a house-hold effect, but it is generally related to the sharing of any non-genetic factor among individuals.
The house-hold effect is included in the analysis if the pedigree data contain a house-hold ID field
(the expected field name is `hhid`).
`SOLAR` automatically creates `Pedigree-Household Groups` based on the `hhid` identifiers.
That means that pedigrees and household groups are merged into larger groups,
which contain every individual in the same pedigree as well as every individual that has the same 
`hhid` as any individual in the pedigree [@SOLARDOC, Section 9.3 [Household Group Analysis](http://helix.nih.gov/Documentation/solar-6.6.2-doc/09.chapter.html#household)]. 
In a consequence of this grouping operation by `SOLAR`, 
the covariance matrix $H$ of the $c$ random effect
contains the information related to the pedigree-household groups
rather than relationships given by the `hhid` field.

The third $e$ term is the residual error with the diagonal covariance matrix.
It is important to note that the `residual` command in `SOLAR`
[@SOLARDOC, Appendix 1, [residual](http://helix.nih.gov/Documentation/solar-6.6.2-doc/91.appendix_1_text.html#residual) command]
returns the covariate-free value of a trait under study, expressed as $y - X \beta$,
and the returned value is not the residual vector $e$ given in the equation of the polygenic model.

The polygenic term $g$ represents the additive genetic effect by default.
The incorporation of the dominance effect into `SOLAR` models is also possible,
but it is attributed to advanced modeling topics [@SOLARDOC, Section 9.4 [Dominance Analysis](http://helix.nih.gov/Documentation/solar-6.6.2-doc/09.chapter.html#dominance)].
The `SOLAR`'s developers suggest that the dominance variance component are unlikely useful in quantitative trait mapping, 
particularly in human genetic analysis, as only bi-lineal relatives contribute to this component. 

`SOLAR` is able to work with discrete traits,
which possible values are coded as two consecutive integers, for example, `0` and `1`
[@SOLARDOC, Appendix 1, [discrete-notes](http://helix.nih.gov/Documentation/solar-6.6.2-doc/91.appendix_1_text.html#discrete-notes)].
`SOLAR` uses a _liability threshold model_ to handle discrete traits [@Duggirala1997],
and more details about implementation and practical notes 
are available in  [@SOLARDOC, Section 9.1 [Discrete Traits](http://helix.nih.gov/Documentation/solar-6.6.2-doc/09.chapter.html)].
`SOLAR` authors recommend to work with quantitative traits whenever it is possible.
The user is able to disable the transformation of discrete traits, if it is necessary.

The likelihood-based statistics, in particular Likelihood Ratio Tests (LRTs),
is used for the inference of the linear mixed models implemented in `SOLAR`.
The following table shows the alternative models used to test the significance of the random effects $g$ and $c$.

| Alternative model   | Free parameters | Restricted parameters  |
|----------|-----------|-------|
| No polygenic effect | $\beta$, $\sigma_c^2$, $\sigma_e^2$      | $\sigma_g^2$ = 0|
| No house-hold effect | $\beta$, $\sigma_g^2$, $\sigma_e^2$     | $\sigma_c^2$ = 0|

Under the assumption of multivariate normality, 
the LRT statistic is asymptotically distributed 
as a $\frac{1}{2}:\frac{1}{2}$ mixture of of a $\chi_1^2$ variable
and a point mass at the origin [@Blangero2000].
This statement is applied to any LRT performed for parameters like $\sigma_g^2$,
which is greater or equal to `0` by definition.

`SOLAR` takes the decision of rejecting the house-hold effect 
by considering two factors, the test results and the value of $\sigma_c^2$ parameter.
If value of $\sigma_c^2$ is non-zero, the house-hold effect is retained in the polygenic model,
nevertheless the effect is not significant.
The user has an option to force the rejection of the house-hold effect if it is necessary.

The tests applied to fixed effects (covariates) are also performed by means of LRTs.
In the table given below the alternative model for the test of significance of `age` covariate is shown.

| Alternative model   | Free parameters | Restricted parameters  |
|----------|-----------|-------|
| No age effect | $\sigma_g^2$, $\sigma_c^2$, $\sigma_e^2$      | $\beta_{age}$ = 0|


### Multivariate polygenic model

In principle, one can estimate the multivariate polygenic model 
by simply performing the univariate analysis in an appropriate way.
Consider two traits $y_1$ and $y_2$ measured in each of $n$ individuals,
assuming that each trait follows the same univariate polygenic analysis.
Then the $2 \times n$ dimensional column vector of observations 
is denoted by the stack of univariate vectors,
and the polygenic model has the matrix form:

$$
\begin{pmatrix}
  y_1 \\
  y_2
\end{pmatrix}_{2n \times 1} =
\begin{pmatrix}
  X_1 & 0 \\
  0 & X_2
\end{pmatrix}_{2n \times nk}
\begin{pmatrix}
  \beta_1 \\
  \beta_2
\end{pmatrix}_{2n \times 1}
+
\begin{pmatrix}
  g_1 \\
  g_2
\end{pmatrix}_{2n \times 1}
+
\begin{pmatrix}
  e_1 \\
  e_2
\end{pmatrix}_{2n \times 1}
$$

$$
G_{2n \times 2n} = C_{2 \times 2} \otimes A_{n \times n} =
\begin{pmatrix}
  c_{11} A & c_{12} A \\
  c_{21} A & c_{21} A
\end{pmatrix} =
\begin{pmatrix}
  \sigma_{1}^2 A & \rho \sigma_{1} \sigma_{2} A \\
  \rho \sigma_{1} \sigma_{2} A & \sigma_{2}^2 A
\end{pmatrix}_{g}
$$

$$
R_{2n \times 2n} = E_{2 \times 2} \otimes I_{n \times n} =
\begin{pmatrix}
  e_{11} I & e_{12} I \\
  e_{21} I & e_{21} I
\end{pmatrix} =
\begin{pmatrix}
  \sigma_{1}^2 I & \rho \sigma_{1} \sigma_{2} I \\
  \rho \sigma_{1} \sigma_{2} I & \sigma_{2}^2 I
\end{pmatrix}_{e}
$$

The model can be generalized to $k$ traits,
and other random effects such as house-hold can be added in the similar way.
The _Kronecker product_ notation $\otimes$ is convinient for denoting the covariance matrices.
More details on the multivariate analysis are given in [@Lynch1998, Chapter 26, Multiple Traits, pp. 774].

The multivariate analysis implemented in `SOLAR` allows for any number of traits,
but the tests for significance of the correlation coefficients are performed only for the bivariate models.
The following table describes the main LRTs available in `SOLAR` for the bivariate analysis.

| Alternative model | Free parameters | Restricted parameters |
---|---|---|
No genetic correlation | $(\sigma_1^2, \sigma_2^2)_g$, $(\sigma_1^2, \sigma_2^2, \rho)_e$ | $\rho_g = 0$ |
Complete pleiotropy  | $(\sigma_1^2, \sigma_2^2)_g$, $(\sigma_1^2, \sigma_2^2, \rho)_e$ | $\rho_g = 1$ |
No environmental correlation | $(\sigma_1^2, \sigma_2^2, \rho)_g$, $(\sigma_1^2, \sigma_2^2)_e$ | $\rho_e = 0$ |

It is important to note that the transformation of discrete traits in the multivariate analysis
is automatically performed by `SOLAR`.

## Linkage model in `SOLAR`

`SOLAR` introduces sequential oligogenic linkage analysis, as stated in its name abbreviation.
Consider the linkage model with a single QTL, which is described by another random effect $q$
in comparison with the polygenic model.

$$y = X\beta + g + c + q + e$$

$$
  \mbox{ where } \mbox{ } \mbox{ } g \sim (0, \sigma_g^2 A) 
  \mbox{ and } \mbox{ } \mbox{ } c \sim (0, \sigma_c^2 H)
  \mbox{ and } \mbox{ } \mbox{ } q \sim (0, \sigma_q^2 M)
  \mbox{ and } \mbox{ } \mbox{ } e \sim (0, \sigma_e^2 I)
$$ 
  
$$ \mbox{ implying } \mbox{ } \mbox{ } y \sim (X\beta, V) = (X\beta, \sigma_g^2 A + \sigma_c^2 I + \sigma_q^2 M + \sigma_e^2)$$

The QTL under study $q$ corresponds to a specific position in the genome and an identity-by-descent (IBD) matrix.
The genome-wide linkage analysis assumes that the highly informative markers, such as micro-satellites, are available in the study,
in order to estimate the two-point or multipoint IBD matrices by `SOLAR` or other specific program like `MERLIN`.
Once the IBD matrices are calculated, the genome-wide scan is conducted by sequentially running the linkage model for each genome position.

The alternative model for testing the significance of a single QTL effect 
is the same as the polygenic model.

Alternative model | Free parameters | Restricted parameters |
---|---|---|
No linkage | $\beta, \sigma_g^2, \sigma_c^2$, $\sigma_e^2$ | $\sigma_q^2$ = 0 |

Traditionally, the logarithm (base 10) of odds (LOD) values are used to score the QTLs.
If the LRT statistics for the linkage model is denoted as $\Lambda$, 
then the correspondent LOD score is equal to $\Lambda / (2 ln(10))$.

`SOLAR` also allows for multi-pass linkage scan,
that means that any further scan can be conditioned to QTLs found in the previous scans.
For example, the second-pass scan is conditioned to the previously found QTL $q_1$ has the form:

$$y = X\beta + g + c + q_1 + q + e$$

The second-pass LRT test now looks like in the table below.

Alternative model | Free parameters | Restricted parameters |
---|---|---|
No linkage | $\beta, \sigma_g^2, \sigma_c^2$, $\sigma_{q_1}^2$, $\sigma_e^2$ | $\sigma_q^2$ = 0 |

The linkage model is extended to deal with many traits or discrete traits in the same way as for the polygenic model.

Additional information related to the linkage modeling in `SOLAR` 
can be found in related publications [@Almasy1998], [@Almasy1997], [@Williams1999],
the `SOLAR` tutorial [@SOLARDOC, Chapter 3 [Tutorial](http://helix.nih.gov/Documentation/solar-6.6.2-doc/03.chapter.html)],
and the `SOLAR` manual, for example, [@SOLARDOC, Apendix 1, [multipoint](http://helix.nih.gov/Documentation/solar-6.6.2-doc/91.appendix_1_text.html#multipoint) command].

## Association model in `SOLAR`

A basic association model in `SOLAR` is extended from the polygenic model
by adding a new covariates `snp`.

$$ y = X\beta + \beta_{snp} * snp + g + c + e$$

$$\mbox{ where } \mbox{ } \mbox{ } g \sim (0, \sigma_g^2 A) 
  \mbox{ and } \mbox{ } \mbox{ } c \sim (0, \sigma_c^2 H)
  \mbox{ and } \mbox{ } \mbox{ } e \sim (0, \sigma_e^2 I)$$ 
  
$$ \mbox{ implying } \mbox{ } \mbox{ } y \sim (X\beta, V) = (X\beta, \sigma_g^2 A + \sigma_c^2 H + \sigma_e^2 I)$$

The genome-wide association scan process SNPs variants under the study according to the given equation one by one.
The p-value of association between trait and SNP is based on LRT test
with the alternative model outlined in the table below.
This alternative model is the same as the polygenic model.

Alternative model | Free parameters | Restricted parameters |
---|---|---|
No association | $\beta, \sigma_g^2, \sigma_c^2$, $\sigma_e^2$ | $\beta_{snp} = 0$ |

More details on implementation of the association analysis in `SOLAR`
can be found in the `SOLAR` manual, for example, [@SOLARDOC, Apendix 1, [mga](http://helix.nih.gov/Documentation/solar-6.6.2-doc/91.appendix_1_text.html#mga) and [snp](http://helix.nih.gov/Documentation/solar-6.6.2-doc/91.appendix_1_text.html#snp) commands].


# Software

## Installation

## `solarius` R package 

```{r inc_solarius}
#library(solarius)
library(devtools)
load_all("../")
```

## Data sets

* GAW10 data distributed within `SOLAR`
* Simulated data `dat30` from `multic` R package [@Andrade2006]
* Simulated data `dat50` from `FFBSKAT` R package
* GAIT1 data 

`dat50` data set is adapted from `example.data` data set in 
[FFBSKAT](http://mga.bionet.nsc.ru/soft/FFBSKAT/) R package. 
The following variables are in the data set.

* `phenodata`: A data.frame with the phenotypes `trait`, `sex` and `age`.
* `genodata`: A matrix of 50 genetic variants (columns) given for 66 individuals (rows).
  The genotypes are coded in the format as `1/1`, `1/2` and `2/2`.
* `genocovdata`: A matrix of covariates derived from the genotype data (additive model).
  The covariates are coded as integers 0, 1 and 2.
* `snpdata`: A data.frame of anotation for genetic variantes. 
   The variables iclude `name` of the variant, the chromose `chrom`, `position` in bp and `gene`.
* `kin`: A kinship matrix for 66 individuals.     


# Examples

## Loading data

## Data exploration

### Plot kinship matrix

```{r kin_dat}
data(dat30)
data(dat50) # phenos in `phenodata`, kinship in `kin`
```

```{r kin_solar, cache = TRUE}
kin2.dat30 <- solarKinship2(dat30)
```

```{r kin_plot, cache = TRUE}
plotKinship2(kin2.dat30)
plotKinship2(2*kin)
```

```{r sub_kin_plot, cache = TRUE}
plotKinship2(kin2.dat30[1:30, 1:30])
```

```{r hist_kin, cache = TRUE}
histKinship2(kin2.dat30)
```

### Plot pedigrees

```{r dat30_ped_2}
plotPed(dat30, 2)
```

### Plot residuals

```{r res_mod}
mod <- solarPolygenic(trait1 ~ age, dat30)
mod
```

```{r res}
r <- residuals(mod)
summary(r)
```

```{r plot_res_demo, cache = TRUE}
plotResQQ(mod)
plotResQQ(mod, labels = TRUE)
plotRes(mod)
```


## Polygenic model (univariate)

```{r M1, cache = TRUE}
M1 <- solarPolygenic(trait1 ~ age + sex, dat30)
M1
```

### Testing covariates

```{r M2, cache = TRUE}
M2 <- solarPolygenic(trait1 ~ age + sex, dat30, covtest = TRUE)
M2
```

```{r M2_cf, cache = TRUE}
M2$cf
```

### Custom kinship matrix

`dat50` data set is used.

```{r M3, cache = TRUE}
M3 <- solarPolygenic(trait ~ 1, phenodata, kinship = kin)
M3
```

The polygenic and sporadic models have the same LRT statistics.

```{r M3_lf, cache = TRUE}
M3$lf
```

## Polygenic model (bivariate)

```{r B1, cache = TRUE}
B1 <- solarPolygenic(trait1 + trait2 ~ 1, dat30)
B1
```

### Testing correlations

```{r B2, cache = TRUE}
B2 <- solarPolygenic(trait1 + trait2 ~ 1, dat30, polygenic.options = "-testrhoe -testrhog")
B2
```

```{r B2_vcf, cache = TRUE}
B2$vcf
```

### Trait-specific covariates

```{r B3, cache = TRUE}
B3 <- solarPolygenic(trait1 + trait2 ~ sex + age(trait2), dat30)
B3
```

```{r tail_mod}
l_ply(tail(B3$solar$files$model$poly.mod, 3), function(x) cat(x, "\n"))

```

## Association model

`dat50` data set is used.

### SNP data by `snpdata` argument

```{r genodata, cache = TRUE}
A1 <- solarAssoc(trait ~ 1, phenodata, snpdata = genodata, kinship = kin)
A1
```

```{r A1_summary}
summary(A1)
```

### SNP data by `snpcovdata` argument

```{r genocovdata, cache = TRUE}
A2 <- solarAssoc(trait ~ 1, phenodata, snpcovdata = genocovdata, kinship = kin)
A2
```


```{r genocovdata2, cache = TRUE}
genocovdata2 <- genocovdata
genocovdata2[genocovdata2 == 2] <- 1.9

A2 <- solarAssoc(trait ~ 1, phenodata, snpcovdata = genocovdata2, kinship = kin)
A2
```


### SNP data by `genocov.files` (single value)
 
```{r genocov, cache = TRUE}
dir <- package.file("extdata", "solarAssoc", package = "solarius")
genocov.files <- file.path(dir, "snp.genocov")
snplists.files <- file.path(dir, c("snp.geno-list1", "snp.geno-list2"))

A3 <- solarAssoc(trait ~ 1, phenodata, 
  genocov.files = genocov.files, snplists.files = snplists.files)
A3
```

### SNP data by `genocov.files` (many values)
 
```{r genocov2, cache = TRUE}
dir <- package.file("extdata", "solarAssoc", package = "solarius")
genocov.files <- file.path(dir, c("snp.genocov1", "snp.genocov2"))
snplists.files <- file.path(dir, c("snp.geno-list1", "snp.geno-list2"))

A4 <- solarAssoc(trait ~ 1, phenodata, 
  genocov.files = genocov.files, snplists.files = snplists.files)
A4
```

### Map information and Manhetten plot

```{r assoc_map, cache = TRUE}
A5 <- solarAssoc(trait ~ 1, phenodata, snpdata = genodata, snpmap = snpdata, kinship = kin)
```

```{r manh_plot, cache = TRUE}
plot(A5)
```

### Parallelization

```{r assoc_par, cache = TRUE}
A6 <- solarAssoc(trait ~ 1, phenodata, snpdata = genodata, kinship = kin, cores = 2)
```

```{r A_cpu}
A1$assoc$tprofile$cputime.sec
A6$assoc$tprofile$cputime.sec
```

## Linkage model

`dat30` data set will be used for the linkage examples.

### Basic linkage model

```{r data_multic_link}
mibddir <- package.file("extdata", "solarOutput", "solarMibdsCsv", package = "solarius")  
list.files(mibddir)
```

```{r L1, cache = TRUE}
L1 <- solarMultipoint(trait1 ~ 1, dat30, mibddir = mibddir)
L1
```

```{r L1_sum}
summary(L1)
```

```{r plot_L1, cache = TRUE}
plot(L1)
```

### Linkage options

```{r L2, cache = TRUE}
L2 <- solarMultipoint(trait1 ~ 1, dat30, mibddir = mibddir, chr = 5, interval = 2, 
  multipoint.settings = "finemap off", multipoint.options = "3")
summary(L2)
```

```{r plot_L2, cache = TRUE}
plot(L2, pass = 2)
```

### Bivariate linkage

```{r L3, cache = TRUE}
L3 <- solarMultipoint(trait1 + trait2 ~ 1, dat30, mibddir = mibddir, chr = 5, 
  interval = 2, multipoint.settings = "finemap off")
summary(L3)
```

```{r L4, cache = TRUE}
L4 <- solarMultipoint(trait1 + trait2 ~ age + sex(trait1), dat30, 
  mibddir = mibddir, chr = 5, interval = 2, multipoint.settings = "finemap off")
summary(L4)
```

### Parallelization

```{r L5, cache = TRUE}
L5 <- solarMultipoint(trait1 ~ 1, dat30, mibddir = mibddir,
  chr = c(2, 5), cores = 2)
```

```{r L1_cpu}
L1$multipoint$tprofile$cputime.sec
L5$multipoint$tprofile$cputime.sec
```

# Miscellaneous material

## `SOLAR` references


A list of references given on the official web page of `SOLAR` 
[http://solar.txbiomedgenetics.org/](http://solar.txbiomedgenetics.org/):

* The main reference for SOLAR, including theoretical explanations of the variance component linkage method and the approximate multipoint IBD calculations in pedigrees, is [@Almasy1998];
* Bivariate Quantitative trait linkage is described in the following papers: [@Almasy1997], [@Williams1999];
* The Liability Threshold model for discrete traits is described in the preceding paper as well as the following one: [@Duggirala1997];
* Gene By Environment Interaction is discussed in: [@Towne1997], [@Blangero2009];
* An examination of LOD adjustment is given in: [@Blangero2000], [@Blangero2001];
* Additional references include: [@Blangero1997], [@Williams2004], [@Williams1999b].


# Refences
